x509/build/cert_rustcrypto.rs

```diff
@@ -0,0 +1,534 @@
+    (1_u16 << 8) | svn as u16
+}
+
+#[derive(Sequence, Debug)]
+struct MultiTcbInfo<'a> {
+    tcb_infos: Vec<TcbInfo<'a>>,
+}
+
+impl AssociatedOid for MultiTcbInfo<'_> {
+    const OID: ObjectIdentifier = ObjectIdentifier::new_unwrap(TCG_MULTI_TCB_INFO_OID);
+}
+
+impl AsExtension for MultiTcbInfo<'_> {
+    fn critical(&self, _subject: &Name, _extensions: &[Extension]) -> bool {
+        true
+    }
+}
+
+#[derive(Sequence, Debug)]
+#[asn1(tag_mode = "IMPLICIT")]
+struct TcbInfo<'a> {
+    #[asn1(context_specific = "0", optional = "true", type = "UTF8String")]
+    vendor: Option<String>,
+
+    #[asn1(context_specific = "1", optional = "true", type = "UTF8String")]
+    model: Option<String>,
+
+    #[asn1(context_specific = "2", optional = "true", type = "UTF8String")]
+    version: Option<String>,
+
+    #[asn1(context_specific = "3", optional = "true", tag_mode = "IMPLICIT")]
+    svn: Option<u32>,
+
+    #[asn1(context_specific = "4", optional = "true", tag_mode = "IMPLICIT")]
+    layer: Option<u64>,
+
+    #[asn1(context_specific = "5", optional = "true", tag_mode = "IMPLICIT")]
+    index: Option<u64>,
+
+    #[asn1(context_specific = "6", optional = "true", tag_mode = "IMPLICIT")]
+    fwids: Option<Vec<Fwid<'a>>>,
+
+    #[asn1(
+        context_specific = "7",
+        optional = "true",
+        type = "BIT STRING",
+        tag_mode = "IMPLICIT"
+    )]
+    flags: Option<&'a [u8]>,
+
+    #[asn1(
+        context_specific = "8",
+        optional = "true",
+        type = "OCTET STRING",
+        tag_mode = "IMPLICIT"
+    )]
+    vendor_info: Option<&'a [u8]>,
+
+    #[asn1(
+        context_specific = "9",
+        optional = "true",
+        type = "OCTET STRING",
+        tag_mode = "IMPLICIT"
+    )]
+    tcb_type: Option<&'a [u8]>,
+
+    #[asn1(
+        context_specific = "10",
+        optional = "true",
+        type = "BIT STRING",
+        tag_mode = "IMPLICIT"
+    )]
+    flags_mask: Option<&'a [u8]>,
+}
+
+impl AssociatedOid for TcbInfo<'_> {
+    const OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("2.23.133.5.4.1");
+}
+
+impl AsExtension for TcbInfo<'_> {
+    fn critical(&self, _subject: &Name, _extensions: &[Extension]) -> bool {
+        true
+    }
+}
+
+/// CSR Tempate Builder
+pub struct CertTemplateBuilder<'a, Key> {
+    basic_constraints: Option<BasicConstraints>,
+    key_usage: Option<KeyUsage>,
+    tcg_ueid: Option<TcgUeid<'a>>,
+    multi_tcb_info: Option<MultiTcbInfo<'a>>,
+    single_tcb_info: Option<TcbInfo<'a>>,
+    params: Vec<CertTemplateParam>,
+    _phantom: PhantomData<Key>,
+}
+
+pub trait BuilderKeys: Sized {
+    type Signature: spki::SignatureBitStringEncoding;
+    fn key_gen() -> Self;
+}
+
+impl BuilderKeys for ml_dsa::KeyPair<MlDsa87> {
+    type Signature = ml_dsa::Signature<MlDsa87>;
+    fn key_gen() -> Self {
+        let mut rng = rand::thread_rng();
+        <MlDsa87 as KeyGen>::key_gen(&mut rng)
+    }
+}
+
+impl<'a, Key> CertTemplateBuilder<'a, Key>
```

Comment:
This is a lot of duplicated code. Can you just factor out the openssl/rustcrypto calls?

Reply:
I'm not sure if there is that much to gain to factor things out? It's basically adding dummy entries either rustcrypto or openssl format and having needles to those.

Reply:
I think at least factoring out the struct definitions would make sense.

Reply:
Alright done.